dataflowengineoss/src/main/scala/io/shiftleft/dataflowengineoss/passes/reachingdef/ReachingDefPass.scala

```diff
@@ -5,24 +5,47 @@ import io.shiftleft.codepropertygraph.generated.nodes._
 import io.shiftleft.codepropertygraph.generated.{EdgeTypes, PropertyNames}
 import io.shiftleft.passes.{DiffGraph, ParallelCpgPass}
 import io.shiftleft.semanticcpg.language._
+import org.slf4j.{Logger, LoggerFactory}
 
 import scala.collection.{Set, mutable}
 
 /**
   * A pass that calculates reaching definitions ("data dependencies").
   * */
-class ReachingDefPass(cpg: Cpg) extends ParallelCpgPass[Method](cpg) {
+class ReachingDefPass(cpg: Cpg, maxNumberOfDefinitions: Int = 1024) extends ParallelCpgPass[Method](cpg) {
 
-  override def partIterator: Iterator[Method] = cpg.method.iterator
+  private val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+  override def partIterator: Iterator[Method] = cpg.method.internal.iterator
 
   override def runOnPart(method: Method): Iterator[DiffGraph] = {
+    logger.info("Calculating reaching definitions for: {} in {}", method.fullName, method.filename)
     val problem = ReachingDefProblem.create(method)
+
+    if (shouldBailOut(problem)) {
+      logger.warn("Bailing out.")
+      return Iterator()
+    }
+
     val solution = new DataFlowSolver().calculateMopSolutionForwards(problem)
     val dstGraph = addReachingDefEdges(method, solution)
     addEdgesFromLoneIdentifiersToExit(dstGraph, method, solution)
     Iterator(dstGraph.build())
   }
 
+  private def shouldBailOut(problem: DataFlowProblem[mutable.Set[Definition]]): Boolean = {
+    val method = problem.flowGraph.entryNode.asInstanceOf[Method]
+    val transferFunction = problem.transferFunction.asInstanceOf[ReachingDefTransferFunction]
+    val numberOfDefinitions = transferFunction.gen.foldLeft(0)(_ + _._2.size)
```

Comment:
Not a blocker, but splitting this line up or adding a comment to explain why this works could potentially be useful for understanding why this works. The variable name is clear enough to convey the intent, however.